### ISSUE TITLE
Wrong colors in `color_list` from `scipy.cluster.hierarchy.dendrogram`

### ISSUE THREAD

**Issue by jolespin** [scipy/scipy#6346]
```
# Init
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns; sns.set()

# Load data
from sklearn.datasets import load_diabetes

# Clustering
from scipy.cluster.hierarchy import dendrogram, fcluster, leaves_list
from scipy.spatial import distance
from fastcluster import linkage # You can use SciPy one too

%matplotlib inline

# Dataset
A_data = load_diabetes().data
DF_diabetes = pd.DataFrame(A_data, columns = ["attr_%d" % j for j in range(A_data.shape[1])])

# Absolute value of correlation matrix, then subtract from 1 for disimilarity
DF_dism = 1 - np.abs(DF_diabetes.corr())

# Compute average linkage
A_dist = distance.squareform(DF_dism.as_matrix())
Z = linkage(A_dist,method="average")
```
The colors specified by the `color_list` key don't sync up with the leaves. 

Can future versions please allow the `link_color_func` to support a dictionary object that has `{key = leaf_label: value=color}` ? It seems to be the most intuitive way for the user to specify colors. 

![screen shot 2016-07-02 at 6 33 29 pm](https://cloud.githubusercontent.com/assets/9061708/16543323/d024b248-4083-11e6-9db3-0997b1babf54.png)

**Comment by jolespin** [comment]
I just realized that `dcoord` and `icoord` have a smaller size than `ivl` and `leaves` ? 

```
icoord	9
ivl	10
color_list	9
leaves	10
dcoord	9
```

**Comment by WarrenWeckesser** [comment]
Here's my answer on stackoverflow:

-----

`icoord`, `dcoord` and `color_list` describe the *links*, not the leaves.  `icoord` and `dcoord` give the coordinates of the "arches" (i.e. upside-down U or J shapes) for each link in a plot, and `color_list` is the color of those arches.  In a full plot, the length of `icoord`, etc., will be one less than the length of `ivl`, as you have observed.

Don't try to line up the `ivl` list with the `icoord`, `dcoord` and `color_list` lists. They are associated with different things.

-----

You said *"The colors specified by the `color_list` key don't sync up with the leaves."*  That's because those are the colors of the links, not the leaves.

**Comment by WarrenWeckesser** [comment]
You also made an enhancement request: *"Can future versions please allow the link_color_func to support a dictionary object that has {key = leaf_label: value=color}"*  If my previous comment made sense, you'll see that this request doesn't make sense, because `link_color_func` is for coloring the *links*, not the leaves.  `dendrogram` doesn't color the leaves.  In a full plot, the leaves are simply the *labels* on the x axis.

I'm closing the issue, but feel free to add more comments here.  If I've misunderstood what you want, we can reopen the issue.

**Comment by jolespin** [comment]
Thanks for you answer.  It makes more sense now.  Is there a tutorial available for  custom coloration of the `links` if one has a dictionary of `leaves to colors`? I'm not sure how to traverse the dendrogram function output to reconstruct the dendrogram with custom colors.

**Comment by WarrenWeckesser** [comment]
I don't know of any tutorials like that.  I also don't fully understand what you want to accomplish.  The lines draw by `dendrogram` (which I called "arches" above) are the graphical representation of a link, and each link is associated with at least two leaves.  What method do you propose to map from your leaf color dictionary to link colors?   For example, suppose the dendrogram results in a link between two leaves that are assigned different colors.  What color should be used to draw the link?

**Comment by jolespin** [comment]
Sorry, I think I was using the terminology loosely.  Once I compute the dendrogram, I use`scipy.cluster.hierarchy.fcluster` to find the clusters after cutting the tree at a certain height.  After I have those clusters, I make a dictionary that assigns a specific color to each of the clusters (and all of their leaves).  I want to be able to color my clusters (w/ leaves and links below distance threshold) with the new colors I assigned.  My reason for this was to have non-redundant colors for samples with lots of clusters.  Very similar to the results from the `color_threshold` but with my more flexibility and scalable.
